File: backend/Infrastructure/Data/Entities/Facility.cs
namespace Infrastructure.Data.Entities;

public class Facility : BaseEntity
{
    public required string Name { get; set; }
    public FacilityType FacilityType { get; set; } = FacilityType.None;
    public bool IsAccessible { get; set; }
    public ICollection<FacilityImage>? Images { get; set; }
    public decimal? Latitude { get; set; }
    public decimal? Longitude { get; set; }
    public string? Location { get; set; }
    public string? Description { get; set; }
    public string? Url { get; set; }

    public ICollection<CityArea>? CityAreas { get; set; }
}

File: app/src/data/types/facility.ts
import { TrailImage } from "./trail";

export interface CreateFacilityRequest {
  name: string;
  facilityType: number;
  isAccessible: boolean;
  latitude: number;
  longitude: number;
}

// Facility kinds. Mirrors the backend's [Flags] FacilityType enum: a facility can
// carry several flags at once (a combined fire pit + shelter is 3 = FirePit | Shelter),
// so test membership with hasFacilityType() (bitwise), never equality.
export const FacilityType = {
  None: 0,
  FirePit: 1,
  Shelter: 2,
  FishingArea: 4,
  SwimmingArea: 8,
  NatureReserve: 16,
} as const;

export interface Facility {
  identifier: string;
  name: string;
  facilityType: number;
  isAccessible: boolean;
  latitude?: number;
  longitude?: number;
  location?: string;
  description?: string;
  url?: string;
}

export function hasFacilityType(facilityType: number, type: (typeof FacilityType)[keyof typeof FacilityType]): boolean {
  return (facilityType & type) !== 0;
}

export interface CityArea {
  identifier: string;
  name: string;
  location: string;
  description?: string;
  imageUrl?: string;
  url?: string;
  facilities: Facility[];
  trails: CityAreaTrail[];
}

export interface CityAreaTrail {
  identifier: string;
  name: string;
  trailLength: number;
  classification: number;
  description?: string;
  averageRating: number;
  image?: TrailImage;
}
